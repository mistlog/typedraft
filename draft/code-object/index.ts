export * from "./export-class";
export * from "./inline-context";
export * from "./local-context";
export * from "./method";
export * from "./module";
