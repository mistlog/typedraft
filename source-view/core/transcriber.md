# Transcriber

We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represent the execution flow.

When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not execute. Code is only meaningful within its' context. That's why we choose another word to differentiate it from compiler.

```typescript
export class Transcriber {
    //
    m_Module: ModuleCode;

    // the 3 types of code we will transform
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<ClassMethod>>;
    m_ContextMap: Map<string, LocalContext>;

    //
    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;

    //
    get m_Code() {
        return this.m_Module.m_Code;
    }
    set m_Code(code: Array<Statement>) {
        this.m_Module.m_Code = code;
    }
}
```

A transcriber just likes a container, it's a collection of plugins. For example, to transform local context, we have ```draft-plugin-local-context```, to add class method to class, we have ```draft-plugin-class```, and ```draft-plugin-dsl``` for DSL extension support.

```typescript
<Transcriber /> +
    function Transcribe(this: Transcriber) {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, { comments: false });
    };
```

# Preprocess

```typescript
export interface ITranscriber {
    Preprocess: () => void;
}
```

When we init a transcriber, we will preprocess code to build some maps for lookup purpose.

```typescript
<Transcriber /> +
    function Preprocess(this: Transcriber) {
        const draft = this.m_Module.ToDraft();
        draft.forEach(each => {
            if (each instanceof ExportClassCode) {
                this.m_ClassMap.set(each.m_Name, each);
            } else if (each instanceof MethodCode) {
                const { class_name, method } = each.ToClassMethod();
                const methods = this.m_MethodMap.get(class_name);
                if (!methods) {
                    this.m_MethodMap.set(class_name, [method]);
                } else {
                    methods.push(method);
                }
            } else if (each instanceof LocalContext) {
                this.m_ContextMap.set(each.m_Name, each);
            }
        });
    };
```

# Utility

Other methods are just utility.

## DSL

```typescript
export interface IDSL {
    Transcribe(block: Array<Statement>): Array<Statement>;
}
```

```typescript
export interface ITranscriber {
    GetDSL: (name: string) => IDSL;
    PrepareDSLs: () => void;
}
```

## Plugin

```typescript
export interface IPlugin {
    Transcribe(): void;
}
```

```typescript
export interface ITranscriber {
    PreparePlugins: () => void;
}
```

## Local Context

```typescript
export interface ITranscriber {
    TraverseLocalContext: (callback: ITraverseLocalContextCallback) => void;
    GetLocalContext: (name: string) => LocalContext;
}
```

```typescript
export type ITraverseLocalContextCallback = (context: LocalContext, name: string) => void;
```

## Class

```typescript
export interface ITranscriber {
    GetClass: (name: string) => ExportClassCode;
    TraverseMethod: (callback: ITraverseMethodCallback) => void;
}
```

```typescript
export type ITraverseMethodCallback = (methods: Array<ClassMethod>, class_name: string) => void;
```

# Implementation

##  DSL

```typescript
<Transcriber /> +
    function AddDSL(this: Transcriber, name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    };
```

```typescript
<Transcriber /> +
    function GetDSL(this: Transcriber, name: string) {
        return this.m_DSLMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function PrepareDSLs(this: Transcriber) {
        this.m_DSLMap.set("match", new PatternMatch());
    };
```

##  Plugin

```typescript
<Transcriber /> +
    function PreparePlugins(this: Transcriber & ITranscriber) {
        /**
         * pay attention to the default order of plugins
         * 1. resolve DSL first
         * 2. then context
         * 3. add methods to class
         * 4. remove redundant code
         */
        this.m_Plugins = [new DSLPlugin(this), new LocalContextPlugin(this), new ClassPlugin(this), new FilterPlugin(this)];
    };
```

```typescript
<Transcriber /> +
    function AddPlugin(plugin: IPlugin) {
        this.m_Plugins.push(plugin);
    };
```

##  Local Context

```typescript
<Transcriber /> +
    function GetLocalContext(this: Transcriber, name: string) {
        return this.m_ContextMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function TraverseLocalContext(this: Transcriber, callback: ITraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    };
```

##  Class

```typescript
<Transcriber /> +
    function GetClass(this: Transcriber, name: string) {
        return this.m_ClassMap.get(name);
    };
```

```typescript
<Transcriber /> +
    function TraverseMethod(this: Transcriber, callback: ITraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    };
```

# Trivial

```typescript
export interface ITranscriber {
    m_Code: Array<Statement>;
}
```

```typescript
<Transcriber /> +
    function constructor(this: Transcriber & ITranscriber, _module: ModuleCode | string) {
        //
        this.m_Module = typeof _module === "string" ? new ModuleCode(_module) : _module;

        //
        this.m_ClassMap = new Map();
        this.m_MethodMap = new Map();
        this.m_ContextMap = new Map();
        this.m_DSLMap = new Map();

        //
        this.PrepareDSLs();
        this.PreparePlugins();

        //
        this.Preprocess();
    };
```