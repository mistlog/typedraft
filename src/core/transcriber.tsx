/**
 * # Transcriber
 * We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represents the execution flow.
 * When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not executed.
 * That's why we choose another word to differentiate it from compiler.
 */
export class Transcriber {
    m_Module: ModuleCode;

    /**
     * the 4 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;
}

/**
 * A transcriber is just a container, it's a collection of plugins.
 * For example, to transform local context, we have `draft-plugin-local-context`, to add class method to class, we have `draft-plugin-class`, and `draft-plugin-dsl` for DSL extension support.
 */
<Transcriber /> +
    function Transcribe(this: Transcriber) {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, { comments: false });
    };

/**
 * # Utility
 * Other methods are just utilities.
 */

/**
 * ## DSL
 */

<Transcriber /> +
    function AddDSL(this: Transcriber, name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    };

<Transcriber /> +
    function GetDSL(this: Transcriber, name: string) {
        return this.m_DSLMap.get(name);
    };

/**
 * ## Context and Class
 */
<Transcriber /> +
    function GetLocalContext(this: Transcriber, name: string) {
        return this.m_ContextMap.get(name);
    };

<Transcriber /> +
    function TraverseLocalContext(this: Transcriber, callback: TraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    };

<Transcriber /> +
    function TraverseInlineContext(this: Transcriber, callback: TraverseInlineContextCallback) {
        this.m_InlineContextMap.forEach(context => callback(context));
    };

<Transcriber /> +
    function GetClass(this: Transcriber, name: string) {
        return this.m_ClassMap.get(name);
    };

<Transcriber /> +
    function TraverseMethod(this: Transcriber, callback: TraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    };

/**
 * # Trivial
 */
<Transcriber /> +
    function constructor(this: Transcriber & ITranscriber, _module: string) {
        this.m_Module = new ModuleCode(_module);

        this.m_ClassMap = new Map<string, ExportClassCode>();
        this.m_MethodMap = new Map<string, Array<MethodCode>>();
        this.m_ContextMap = new Map<string, LocalContext>();
        this.m_InlineContextMap = new Map<Symbol, InlineContext>();
        this.m_DSLMap = new Map<string, IDSL>();
        this.m_Plugins = new Array<IPlugin>();
    };

export interface IDSL {
    Transcribe?(
        block: Array<Statement>,
        path?: NodePath<FunctionDeclaration> | NodePath<BlockStatement>
    ): Array<Statement>;

    /**
     * @param literal TODO: NodePath<TemplateLiteral>
     */
    InplaceTranscribe?(literal: any, type: string): CallExpression;

    m_Merge?: boolean;
}

export interface IPlugin {
    Transcribe(): void;
}

export interface ITranscriber {
    Transcribe: () => string;

    AddDSL: (name: string, dsl: IDSL) => void;

    GetDSL: (name: string) => IDSL;
    GetClass: (name: string) => ExportClassCode;
    GetLocalContext: (name: string) => LocalContext;

    TraverseMethod: (callback: TraverseMethodCallback) => void;
    TraverseLocalContext: (callback: TraverseLocalContextCallback) => void;
    TraverseInlineContext: (callback: TraverseInlineContextCallback) => void;

    m_Module: ModuleCode;

    /**
     * the 3 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;
}

export type TraverseLocalContextCallback = (context: LocalContext, name: string) => void;
export type TraverseInlineContextCallback = (context: InlineContext) => void;
export type TraverseMethodCallback = (methods: Array<MethodCode>, class_name: string) => void;

import {
    Statement,
    FunctionDeclaration,
    BlockStatement,
    CallExpression,
    TemplateLiteral,
} from "@babel/types";
import { NodePath } from "@babel/traverse";

import { ToString } from "../common/utility";

import { ModuleCode } from "../code-object/module";
import { ExportClassCode } from "../code-object/export-class";
import { MethodCode } from "../code-object/method";
import { LocalContext } from "../code-object/local-context";
import { InlineContext } from "../code-object/inline-context";
