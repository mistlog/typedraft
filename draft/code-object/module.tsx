import {
    Program,
    File,
    ExpressionStatement,
    ExportNamedDeclaration,
    FunctionDeclaration,
} from "@babel/types";
import { ExportClassCode } from "./export-class";
import { MethodCode } from "./method";
import { LocalContext } from "./local-context";
import { ToFile, ToString } from "../common/utility";
import traverse, { NodePath, Node, TraverseOptions } from "@babel/traverse";
import { InlineContext } from "./inline-context";
import { MatchDSL, __ } from "draft-dsl-match";

/**
 * # Draft
 * A .tsx file is considered as a module and we will transform these 4 types of code, they are collected and then transcibed to "real" code.
 */
export class ModuleCode {
    m_File: File;
    m_Path: NodePath<Program>;

    constructor(code: string) {
        this.m_File = ToFile(code);
    }

    /**
     * As we are only interested in the draft part of a module, then we need a way to return this "view" of module code.
     */
    ToDraft() {
        /**
         * refresh and update bindings because DSL only transforms code
         */
        this.m_File = ToFile(ToString(this.m_File));

        /**
         * traverse file and set path
         */
        let draft: Draft = [];
        this.Traverse<{ _module: ModuleCode }>(
            {
                Program(path) {
                    this._module.m_Path = path;
                },

                ExpressionStatement(path) {
                    const expression = path.get("expression");
                    if (expression.isStringLiteral()) {
                        const literal = expression.node.value.trim();
                        if (literal.startsWith("use") && path.parentPath.isBlockStatement()) {
                            /**
                             * inline context should be resolved from deepest
                             */
                            draft.unshift(new InlineContext(path.parentPath));
                        }
                    }
                },
            },
            { _module: this }
        );

        /**
         * collect draft parts
         */
        this.m_Path.get("body").forEach(path => {
            /**
             * remove redundant ; before tag, see comment of `ToFile` in utility.ts
             */
            if (path.isEmptyStatement()) {
                path.remove();
                return;
            }

            // prettier-ignore
            const to_add = Î›<ExportClassCode | LocalContext | MethodCode | null>("match")` ${path as NodePath<any>} 
                ${IsExportClassCode(path)} -> ${(path: NodePath<ExportNamedDeclaration>) => new ExportClassCode(path)}
                ${IsMethodCode(path)} -> ${(path: NodePath<ExpressionStatement>) => new MethodCode(path)}
                ${IsLocalContext(path)} -> ${(path: NodePath<FunctionDeclaration>) => new LocalContext(path.scope.parent.getBinding(path.node.id.name))}
                ${__} -> ${null}
            `;

            if (to_add) {
                draft.push(to_add);
            }
        });
        return draft;
    }

    Traverse<S>(this: ModuleCode, visitor: TraverseOptions<S>, state: S = null) {
        this.m_File = ToFile(ToString(this.m_File));
        traverse<S>(this.m_File, visitor, null, state);
    }
}

export type Draft = Array<ExportClassCode | MethodCode | LocalContext | InlineContext>;

export function IsExportClassCode(path: NodePath<any>) {
    return (
        path.isExportNamedDeclaration() &&
        (path.get("declaration") as NodePath<Node>).isClassDeclaration()
    );
}

export function IsMethodCode(path: NodePath<any>) {
    if (!path.isExpressionStatement()) {
        return false;
    }

    const expression = path.get("expression") as NodePath<Node>;
    if (!expression.isBinaryExpression()) {
        return false;
    }

    const left = expression.get("left") as NodePath<Node>;
    const right = expression.get("right") as NodePath<Node>;
    return left.isJSXElement() && right.isFunctionExpression();
}

export function IsLocalContext(path: NodePath<any>) {
    if (!path.isFunctionDeclaration()) {
        return false;
    }

    const [directive]: [string] = path.node.body.directives;

    const is_local_context = path.scope.parent
        .getBinding(path.node.id.name)
        .referencePaths.some(path => {
            const used_as_jsx = path.parentPath?.parentPath?.isJSXElement();
            const used_as_statement = path.parentPath?.parentPath?.parentPath?.isExpressionStatement();
            const has_context = Boolean(directive);
            return used_as_jsx && (has_context || used_as_statement);
        });
    return is_local_context;
}
