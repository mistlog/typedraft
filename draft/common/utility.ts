import { template } from "@babel/core";
import { Node } from "@babel/types";
import { NodePath } from "@babel/traverse";
import { default as Binding } from "@babel/traverse/lib/scope/binding";
import { parse } from "@babel/parser";
import generate, { GeneratorOptions } from "@babel/generator";

export function ToAst<T = any>(code: string) {
    return (template.ast(code, { plugins: ["typescript", "jsx"] }) as unknown) as T;
}

export function ToNodePath<T = any>(code: string): NodePath<T> {
    const path = new NodePath<T>(null, null);
    path.node = (ToAst(code) as unknown) as T;
    return path;
}

/**
 * only used in test to construct required binding param
 */
export function ToBinding(code: string): Binding {
    const binding = new Binding({
        path: ToNodePath<Node>(code),
        existing: null,
        identifier: null,
        scope: null,
        kind: "let",
    });
    return binding;
}

export function ToFile(raw: string) {
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<", "gm"), ";<");
    return parse(code, {
        sourceType: "module",
        plugins: ["typescript", "jsx", ["decorators", { decoratorsBeforeExport: true }]],
    });
}

export function ToString(node: Node, options?: GeneratorOptions) {
    return generate(node, options || {}).code;
}
