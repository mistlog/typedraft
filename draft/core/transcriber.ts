import {
    Statement,
    FunctionDeclaration,
    BlockStatement,
    Expression,
    TemplateLiteral,
} from "@babel/types";
import { NodePath } from "@babel/traverse";
import { ToString } from "../common/utility";
import { ModuleCode } from "../code-object/module";
import { ExportClassCode } from "../code-object/export-class";
import { MethodCode } from "../code-object/method";
import { LocalContext } from "../code-object/local-context";
import { InlineContext } from "../code-object/inline-context";

/**
 * # Transcriber
 * We always write code in one language and translate(compile) it to another, however, in both language, the code we write directly represents the execution flow.
 * When we write draft, we don't expect it executes as is, instead, it's designed to be transformed, not executed.
 * That's why we choose another word to differentiate it from compiler.
 */
export class Transcriber {
    m_Module: ModuleCode;

    /**
     * the 4 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;

    constructor(_module: string) {
        this.m_Module = new ModuleCode(_module);

        this.m_ClassMap = new Map<string, ExportClassCode>();
        this.m_MethodMap = new Map<string, Array<MethodCode>>();
        this.m_ContextMap = new Map<string, LocalContext>();
        this.m_InlineContextMap = new Map<Symbol, InlineContext>();
        this.m_DSLMap = new Map<string, IDSL>();
        this.m_Plugins = new Array<IPlugin>();
    }

    /**
     * A transcriber is just a container, it's a collection of plugins.
     * For example, to transform local context, we have `draft-plugin-local-context`, to add class method to class, we have `draft-plugin-class`, and `draft-plugin-dsl` for DSL extension support.
     */

    Transcribe() {
        this.m_Plugins.forEach(plugin => plugin.Transcribe());
        return ToString(this.m_Module.m_File, {
            comments: false,
        });
    }

    AddDSL(name: string, dsl: IDSL) {
        this.m_DSLMap.set(name, dsl);
    }

    GetDSL(name: string) {
        return this.m_DSLMap.get(name);
    }

    GetLocalContext(name: string) {
        return this.m_ContextMap.get(name);
    }

    TraverseLocalContext(callback: TraverseLocalContextCallback) {
        this.m_ContextMap.forEach((context, name) => callback(context, name));
    }

    TraverseInlineContext(callback: TraverseInlineContextCallback) {
        this.m_InlineContextMap.forEach(context => callback(context));
    }

    GetClass(name: string) {
        return this.m_ClassMap.get(name);
    }

    TraverseMethod(callback: TraverseMethodCallback) {
        this.m_MethodMap.forEach((methods, class_name) => callback(methods, class_name));
    }
}

export interface IDSL {
    Transcribe?(
        block: Array<Statement>,
        path?: NodePath<FunctionDeclaration> | NodePath<BlockStatement>
    ): Array<Statement>;

    InplaceTranscribe?(literal: NodePath<TemplateLiteral>, type: string): Expression;

    m_Merge?: boolean;
}

export interface IPlugin {
    Transcribe(): void;
}

export interface ITranscriber {
    Transcribe: () => string;

    AddDSL: (name: string, dsl: IDSL) => void;

    GetDSL: (name: string) => IDSL;
    GetClass: (name: string) => ExportClassCode;
    GetLocalContext: (name: string) => LocalContext;

    TraverseMethod: (callback: TraverseMethodCallback) => void;
    TraverseLocalContext: (callback: TraverseLocalContextCallback) => void;
    TraverseInlineContext: (callback: TraverseInlineContextCallback) => void;

    m_Module: ModuleCode;

    /**
     * the 3 types of code we will transform
     */
    m_ClassMap: Map<string, ExportClassCode>;
    m_MethodMap: Map<string, Array<MethodCode>>;
    m_ContextMap: Map<string, LocalContext>;
    m_InlineContextMap: Map<Symbol, InlineContext>;

    m_DSLMap: Map<string, IDSL>;
    m_Plugins: Array<IPlugin>;
}

export type TraverseLocalContextCallback = (context: LocalContext, name: string) => void;
export type TraverseInlineContextCallback = (context: InlineContext) => void;
export type TraverseMethodCallback = (methods: Array<MethodCode>, class_name: string) => void;
