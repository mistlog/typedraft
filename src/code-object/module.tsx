import { Program, File, ExpressionStatement, ExportNamedDeclaration, FunctionDeclaration } from "@babel/types";
import { ExportClassCode } from "./export-class";
import { MethodCode } from "./method";
import { LocalContext } from "./local-context";
import { Visitor } from "@babel/core";
import { ToFile } from "../common/utility";
import traverse, { NodePath, Node } from "@babel/traverse";

/*
# Draft

A .tsx file is considered as a module and we will transform these 3 types of code, they are collected and then transcibed to "real" code.

*/
export type Draft = Array<ExportClassCode | MethodCode | LocalContext>;

export class ModuleCode
{
    m_File: File;
    m_Path: NodePath<Program>;
}

/*
## Example

Typical file with draft would be:

```typescript
export class Foo {
    static foo: number;
}

<Foo/> + function Test(this: Foo, a: number, b: string){
    <Bar/>;
    return a.toString()+b;
};

function Bar(this: Foo, a: number, b: string){
    a += this.foo;
    <Nested/>
    console.log("bar");
}

function Nested(){
    console.log("nested");
}
```

The transcribed code would be:

```typescript
export class Foo {
  static foo: number;

  Test(a: number, b: string) {
    a += this.foo;
    console.log("nested");
    console.log("bar");
    return a.toString() + b;
  }
}
```

*/

/*
# View Module as Draft

As we are only interested in the draft part of a module, then we need a way to return this "view" of module code.
*/
<ModuleCode /> + function ToDraft(this: ModuleCode)
{
    let draft: Draft = [];

    const visitor: Visitor = {
        Program(path)
        {
            path.get("body").forEach(path =>
            {
                //@ts-ignore
                <CreateDraft />;
            })
        }
    }

    traverse(this.m_File, visitor);
    return draft;
};

/*
## Create draft
*/

function CreateDraft(path: NodePath<Node>, draft: Draft)
{
    if (IsExportClassCode(path))
    {
        draft.push(new ExportClassCode(path.node, path));
    }
    else if (IsMethodCode(path))
    {
        draft.push(new MethodCode(path.node, path));
    }
    else if (IsLocalContext(path))
    {
        const name = path.node.id.name;
        const binding = path.scope.parent.getBinding(name);
        draft.push(new LocalContext(path.node, binding, path));
    }
}

export function IsExportClassCode(path: NodePath<Node>): path is NodePath<ExportNamedDeclaration>
{
    if (!path.isExportNamedDeclaration())
    {
        return false;
    }

    const is_export_class = path.get("declaration").isClassDeclaration();
    return is_export_class;
}

export function IsMethodCode(path: NodePath<Node>): path is NodePath<ExpressionStatement>
{
    if (!path.isExpressionStatement())
    {
        return false;
    }

    const expression = path.get("expression");
    if (!expression.isBinaryExpression())
    {
        return false;
    }

    const left_is_jsx = expression.get("left").isJSXElement();
    const right_is_function = expression.get("right").isFunctionExpression();
    return left_is_jsx && right_is_function;
}

export function IsLocalContext(path: NodePath<Node>): path is NodePath<FunctionDeclaration>
{
    if (!path.isFunctionDeclaration())
    {
        return false;
    }

    const name = path.node.id.name;
    const binding = path.scope.parent.getBinding(name);
    const is_local_context = binding.referencePaths.some(path =>
    {
        const used_as_jsx = path.parentPath?.parentPath?.isJSXElement();
        const used_as_statement = path.parentPath?.parentPath?.parentPath?.isExpressionStatement();
        return used_as_jsx && used_as_statement;
    });
    return is_local_context;
}
/*
You may want to refer to the usage of [babel](https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#toc-bindings).
*/

/*
# Trivial
*/
<ModuleCode /> + function constructor(this: ModuleCode, code: string)
{
    this.m_File = ToFile(code);
};

