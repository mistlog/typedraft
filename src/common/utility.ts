import { template } from "@babel/core";
import { Node } from "@babel/types";
import generate, { GeneratorOptions } from "@babel/generator";
import { parse } from "@babel/parser";

export function ToAst(code: string)
{
    return template.ast(code, { plugins: ["typescript", "jsx"] });
}

export function ToFile(raw: string)
{
    // babel bug: https://github.com/babel/babel/issues/8837
    // refert to test case "transcriber.interface-no-parse-error"
    const code = raw.replace(new RegExp("^<","gm"),";<");
    return parse(code, { sourceType: "module", plugins: ["typescript", "jsx", "optionalChaining"] });
}

export function ToString(node: Node, options?: GeneratorOptions)
{
    return generate(node, options || {}).code;
}