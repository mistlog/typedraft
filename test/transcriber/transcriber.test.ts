import { ModuleCode } from "../../src/code-object/module";
import { ToString } from "../../src/common/utility";
import { Transcriber } from "../../src/core/transcriber";

describe("transcriber", () =>
{
    test("transcriber.preprocess", () =>
    {
        //
        const code = `
            export class Foo{
                static foo: number;
            }

            <Foo/> + function Test(this: Foo, a: number, b: string){
                return a.toString()+b;
            }

            function Bar(this: Foo, a: number, b: string){
                a += this.foo;
                return a.toString()+b;
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.Preprocess();

        expect(transcriber.m_ClassMap.size).toEqual(1);
        expect(transcriber.m_MethodMap.size).toEqual(1);
        expect(transcriber.m_ContextMap.size).toEqual(1);
    })

    test("transcriber.transcribe", () =>
    {
        //
        const code = `
            export class Foo{
                static foo: number;
            }

            <Foo/> + function Test(this: Foo, a: number, b: string){
                return a.toString()+b;
            };

            <Foo/> + function Test2(this: Foo, a: number, b: string){
                return a.toString()+b;
            }
        `;

        const transcriber = new Transcriber(code);
        transcriber.Transcribe();
        const result = transcriber.m_ClassMap.get("Foo").m_Code;
        expect(ToString(result)).toMatchSnapshot();
    })

    test("transcriber.interface-no-parse-error", () =>
    {
        //
        const code = `
            interface IFoo {

                // ; is not required after interface and before tag
            }

            <Foo/> + function Test(this: Foo, a: number, b: string){
                return a.toString()+b;
            }
        `;

        expect(() => new Transcriber(code)).not.toThrow();
    })
})