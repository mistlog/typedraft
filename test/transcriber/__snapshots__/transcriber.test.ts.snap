// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transcriber transcriber.no-empty-statement 1`] = `
"export class Foo {
  static foo: number;

  Test(a: number, b: string) {
    a += Foo.foo;
    return a.toString() + b;
  }

}"
`;

exports[`transcriber transcriber.transcribe 1`] = `
"export class Foo {
  static foo: number;

  constructor() {
    console.log(\\"hi\\");
  }

  Test(a: number, b: string) {
    return a.toString() + b;
  }

  Test2(a: number, b: string) {
    return a.toString() + b;
  }

}"
`;
